core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -248,7 +253,7 @@ public ShardFailedClusterStateTaskExecutor(AllocationService allocationService,
             BatchResult.Builder<ShardEntry> batchResultBuilder = BatchResult.builder();
             List<ShardEntry> tasksToBeApplied = new ArrayList<>();
             List<FailedRerouteAllocation.FailedShard> shardRoutingsToBeApplied = new ArrayList<>();
-            Set<ShardRouting> seenShardRoutings = new HashSet<>(); // to prevent duplicates
+            List<FailedRerouteAllocation.StaleShard> staleShardsToBeApplied = new ArrayList<>();
```

Comment:
Before #19715, this duplicate removal logic did not exist. It was the responsibility of `AllocationService.applyFailedShards` to deal with this. As part of the mentioned PR, I changed the contract of `AllocationService.applyFailedShards` and `AllocationService.applyStartedShards` to allow as parameters only shard routing objects that exist in the routing table, and no duplicates. The reason for this was so that these methods would not need the logic to check if a `ShardRouting` object actually existed like that in the routing table (or get the latest routing entry), simplifying its implementation considerably. For `AllocationService.applyFailedShards` this turned out to be very tricky though. In 180eff14ddd022fbb3c92392ec54156ac62b64e1 I reverted some of the behavior with the following explanation:

> Fix issue when relocation source and target routings are failed in same batch update  PR #19715 made AllocationService less lenient, requiring ShardRouting instances that are passed to its applyStartedShards and applyFailedShards methods to exist in the routing table. As primary shard failures also fail initializing replica shards, concurrent replica shard failures that are treated in the same cluster state update might not reference existing replica entries in the routing table anymore. To solve this, PR #19715 ordered the failures by first handling replica before primary failures. There are other failures that influence more than one routing entry, however. When we have a failed shard entry for both a relocation source and target, then, depending on the order, either one or the other might point to an out-dated shard entry. As finding a good order is more difficult than applying the failures, this commit re-adds parts of the ShardRouting re-resolve logic so that the applyFailedShards method can properly treat shard failure batches.

With this re-resolve logic in place, we don't actually need to care about duplicates anymore, as duplicate occurrences will just be ignored (see https://github.com/ywelsch/elasticsearch/blob/ff17552b8d561232d7bfd51ea3ca8730d195fbe1/core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L177)